File: src/js/services/websocket_service.js
/**
 * WebSocket Service for Real-time Updates
 * Manages WebSocket connections and handles real-time training updates
 */

class WebSocketService {
    constructor() {
        this.ws = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 5;
        this.reconnectDelay = 1000;
        this.isConnected = false;
        this.messageQueue = [];
        this.subscriptions = new Map(); // job_id -> callback functions
        this.eventHandlers = new Map();
        this.pingInterval = null;
        this.connectionPromise = null;
    }
    
    /**
     * Connect to WebSocket server
     */
    async connect() {
        if (this.isConnected || this.connectionPromise) {
            return this.connectionPromise;
        }
        
        this.connectionPromise = new Promise((resolve, reject) => {
            const token = localStorage.getItem('token');
            if (!token) {
                reject(new Error('No authentication token found'));
                return;
            }
            
            // Determine WebSocket URL based on current location
            const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
            const host = window.location.hostname;
            const port = window.location.port || (protocol === 'wss:' ? '443' : '80');
            const wsUrl = `${protocol}//${host}:${port}/ws?token=${token}`;
            
            try {
                this.ws = new WebSocket(wsUrl);
                
                this.ws.onopen = () => {
                    console.log('WebSocket connected');
                    this.isConnected = true;
                    this.reconnectAttempts = 0;
                    
                    // Process queued messages
                    this.processMessageQueue();
                    
                    // Start ping interval
                    this.startPingInterval();
                    
                    // Trigger connected event
                    this.triggerEvent('connected');
                    
                    resolve();
                };
                
                this.ws.onmessage = (event) => {
                    this.handleMessage(event.data);
                };
                
                this.ws.onerror = (error) => {
                    console.error('WebSocket error:', error);
                    this.triggerEvent('error', error);
                };
                
                this.ws.onclose = (event) => {
                    console.log('WebSocket disconnected:', event.code, event.reason);
                    this.isConnected = false;
                    this.connectionPromise = null;
                    
                    // Clear ping interval
                    this.stopPingInterval();
                    
                    // Trigger disconnected event
                    this.triggerEvent('disconnected', event);
                    
                    // Attempt reconnection
                    if (this.shouldReconnect(event.code)) {
                        this.reconnect();
                    }
                    
                    reject(new Error(`WebSocket closed: ${event.reason}`));
                };
            } catch (error) {
                console.error('Failed to create WebSocket:', error);
                this.connectionPromise = null;
                reject(error);
            }
        });
        
        return this.connectionPromise;
    }
    
    /**
     * Disconnect from WebSocket server
     */
    disconnect() {
        if (this.ws) {
            this.isConnected = false;
            this.stopPingInterval();
            this.ws.close();
            this.ws = null;
        }
    }
    
    /**
     * Reconnect to WebSocket server
     */
    async reconnect() {
        if (this.reconnectAttempts >= this.maxReconnectAttempts) {
            console.error('Max reconnection attempts reached');
            this.triggerEvent('reconnect_failed');
            return;
        }
        
        this.reconnectAttempts++;
        const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1); // Exponential backoff
        
        console.log(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts}/${this.maxReconnectAttempts})`);
        
        setTimeout(async () => {
            try {
                await this.connect();
                
                // Re-subscribe to all active jobs
                for (const jobId of this.subscriptions.keys()) {
                    await this.subscribeToJob(jobId);
                }
            } catch (error) {
                console.error('Reconnection failed:', error);
            }
        }, delay);
    }
    
    /**
     * Check if should attempt reconnection based on close code
     */
    shouldReconnect(code) {
        // Don't reconnect for authentication errors or normal closure
        const noReconnectCodes = [1000, 1001, 1008];
        return !noReconnectCodes.includes(code);
    }
    
    /**
     * Send a message to the server
     */
    send(message) {
        if (this.isConnected && this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify(message));
        } else {
            // Queue message for later
            this.messageQueue.push(message);
            
            // Try to connect if not connected
            if (!this.isConnected) {
                this.connect().catch(console.error);
            }
        }
    }
    
    /**
     * Process queued messages
     */
    processMessageQueue() {
        while (this.messageQueue.length > 0 && this.isConnected) {
            const message = this.messageQueue.shift();
            this.send(message);
        }
    }
    
    /**
     * Handle incoming messages
     */
    handleMessage(data) {
        try {
            const message = JSON.parse(data);
            
            // Handle ping/pong
            if (message.type === 'ping') {
                this.send({ type: 'pong' });
                return;
            }
            
            // Log message for debugging
            console.log('WebSocket message received:', message);
            
            // Trigger event handlers
            this.triggerEvent('message', message);
            
            // Handle specific message types
            switch (message.type) {
                case 'training_progress':
                    this.handleTrainingProgress(message.data);
                    break;
                
                case 'training_complete':
                    this.handleTrainingComplete(message.data);
                    break;
                
                case 'training_failed':
                    this.handleTrainingFailed(message.data);
                    break;
                
                case 'validation_progress':
                    this.handleValidationProgress(message.data);
                    break;
                
                case 'model_update':
                    this.handleModelUpdate(message.data);
                    break;
                
                case 'connection':
                    this.handleConnectionMessage(message.data);
                    break;
                
                default:
                    console.warn('Unknown message type:', message.type);
            }
        } catch (error) {
            console.error('Error handling WebSocket message:', error);
        }
    }
    
    /**
     * Subscribe to training job updates
     */
    async subscribeToJob(jobId, callbacks = {}) {
        // Store callbacks
        this.subscriptions.set(jobId, callbacks);
        
        // Send subscription message
        this.send({
            type: 'subscribe',
            job_id: jobId
        });
        
        console.log(`Subscribed to job ${jobId}`);
    }
    
    /**
     * Unsubscribe from training job updates
     */
    unsubscribeFromJob(jobId) {
        // Remove callbacks
        this.subscriptions.delete(jobId);
        
        // Send unsubscribe message
        this.send({
            type: 'unsubscribe',
            job_id: jobId
        });
        
        console.log(`Unsubscribed from job ${jobId}`);
    }
    
    /**
     * Handle training progress updates
     */
    handleTrainingProgress(data) {
        const jobId = data.job_id;
        const callbacks = this.subscriptions.get(jobId);
        
        if (callbacks && callbacks.onProgress) {
            callbacks.onProgress(data);
        }
        
        // Trigger global event
        this.triggerEvent('training_progress', data);
        
        // Update UI elements if they exist
        this.updateProgressUI(data);
    }
    
    /**
     * Handle training completion
     */
    handleTrainingComplete(data) {
        const jobId = data.job_id;
        const callbacks = this.subscriptions.get(jobId);
        
        if (callbacks && callbacks.onComplete) {
            callbacks.onComplete(data);
        }
        
        // Trigger global event
        this.triggerEvent('training_complete', data);
        
        // Unsubscribe from job
        this.unsubscribeFromJob(jobId);
        
        // Show completion notification
        this.showNotification('Training Complete', `Model training completed successfully!`, 'success');
    }
    
    /**
     * Handle training failure
     */
    handleTrainingFailed(data) {
        const jobId = data.job_id;
        const callbacks = this.subscriptions.get(jobId);
        
        if (callbacks && callbacks.onError) {
            callbacks.onError(data);
        }
        
        // Trigger global event
        this.triggerEvent('training_failed', data);
        
        // Unsubscribe from job
        this.unsubscribeFromJob(jobId);
        
        // Show error notification
        this.showNotification('Training Failed', data.error_message || 'Model training failed', 'error');
    }
    
    /**
     * Handle validation progress updates
     */
    handleValidationProgress(data) {
        this.triggerEvent('validation_progress', data);
    }
    
    /**
     * Handle model updates
     */
    handleModelUpdate(data) {
        this.triggerEvent('model_update', data);
    }
    
    /**
     * Handle connection messages
     */
    handleConnectionMessage(data) {
        console.log('Connection message:', data);
        
        if (data.status === 'subscribed') {
            console.log(`Successfully subscribed to job ${data.job_id}`);
        } else if (data.status === 'unsubscribed') {
            console.log(`Successfully unsubscribed from job ${data.job_id}`);
        }
    }
    
    /**
     * Update progress UI elements
     */
    updateProgressUI(data) {
        // Update progress modal if it exists
        const progressFill = document.getElementById('generation-progress-fill');
        if (progressFill) {
            progressFill.style.width = data.progress_percentage + '%';
            
            const progressPercentage = progressFill.parentElement.querySelector('.progress-percentage');
            if (progressPercentage) {
                progressPercentage.textContent = Math.round(data.progress_percentage) + '%';
            }
        }
        
        // Update status message
        const statusMessage = document.getElementById('status-message');
        if (statusMessage) {
            statusMessage.textContent = data.message;
        }
        
        // Update stage indicators
        const stages = document.querySelectorAll('.stage-item');
        stages.forEach(stage => {
            const stageName = stage.dataset.stage;
            if (stageName === data.current_stage) {
                stage.classList.add('active');
                stage.classList.remove('completed');
            } else if (this.isStageCompleted(stageName, data.current_stage)) {
                stage.classList.add('completed');
                stage.classList.remove('active');
            } else {
                stage.classList.remove('active', 'completed');
            }
        });
        
        // Update metrics if available
        if (data.metrics) {
            this.updateMetricsDisplay(data.metrics);
        }
    }
    
    /**
     * Check if a stage is completed based on current stage
     */
    isStageCompleted(stageName, currentStage) {
        const stageOrder = ['initializing', 'preprocessing', 'training', 'evaluating', 'finalizing'];
        const stageIndex = stageOrder.indexOf(stageName);
        const currentIndex = stageOrder.indexOf(currentStage);
        return stageIndex < currentIndex;
    }
    
    /**
     * Update metrics display
     */
    updateMetricsDisplay(metrics) {
        // Update accuracy if available
        if (metrics.accuracy !== undefined) {
            const accuracyElement = document.getElementById('current-accuracy');
            if (accuracyElement) {
                accuracyElement.textContent = (metrics.accuracy * 100).toFixed(1) + '%';
            }
        }
        
        // Update other metrics
        for (const [key, value] of Object.entries(metrics)) {
            const element = document.getElementById(`metric-${key}`);
            if (element) {
                if (typeof value === 'number') {
                    element.textContent = value.toFixed(3);
                } else {
                    element.textContent = value;
                }
            }
        }
    }
    
    /**
     * Show notification
     */
    showNotification(title, message, type = 'info') {
        // Check if we have a notification system
        if (window.showToast) {
            window.showToast(message, type);
        } else {
            // Fallback to console
            console.log(`[${type.toUpperCase()}] ${title}: ${message}`);
        }
    }
    
    /**
     * Register event handler
     */
    on(event, handler) {
        if (!this.eventHandlers.has(event)) {
            this.eventHandlers.set(event, []);
        }
        this.eventHandlers.get(event).push(handler);
    }
    
    /**
     * Remove event handler
     */
    off(event, handler) {
        if (this.eventHandlers.has(event)) {
            const handlers = this.eventHandlers.get(event);
            const index = handlers.indexOf(handler);
            if (index > -1) {
                handlers.splice(index, 1);
            }
        }
    }
    
    /**
     * Trigger event
     */
    triggerEvent(event, data = null) {
        if (this.eventHandlers.has(event)) {
            const handlers = this.eventHandlers.get(event);
            handlers.forEach(handler => {
                try {
                    handler(data);
                } catch (error) {
                    console.error(`Error in event handler for ${event}:`, error);
                }
            });
        }
    }
    
    /**
     * Start ping interval to keep connection alive
     */
    startPingInterval() {
        this.stopPingInterval();
        this.pingInterval = setInterval(() => {
            if (this.isConnected) {
                // Server will send ping, we respond with pong
            }
        }, 30000); // Every 30 seconds
    }
    
    /**
     * Stop ping interval
     */
    stopPingInterval() {
        if (this.pingInterval) {
            clearInterval(this.pingInterval);
            this.pingInterval = null;
        }
    }
    
    /**
     * Get connection status
     */
    getStatus() {
        return {
            isConnected: this.isConnected,
            reconnectAttempts: this.reconnectAttempts,
            activeSubscriptions: Array.from(this.subscriptions.keys())
        };
    }
}

// Create singleton instance
const wsService = new WebSocketService();

// Auto-connect when page loads if user is authenticated
if (localStorage.getItem('token')) {
    wsService.connect().catch(console.error);
}

export default wsService;

File: backend/services/websocket_service.py
"""
WebSocket Service for Real-time Training Updates
Manages WebSocket connections and broadcasts training progress
"""

from fastapi import WebSocket, WebSocketDisconnect
from typing import Dict, List, Set, Any, Optional
import json
import asyncio
import logging
from datetime import datetime
from dataclasses import dataclass, asdict
from enum import Enum

logger = logging.getLogger(__name__)

class MessageType(Enum):
    CONNECTION = "connection"
    TRAINING_PROGRESS = "training_progress"
    TRAINING_COMPLETE = "training_complete"
    TRAINING_FAILED = "training_failed"
    MODEL_UPDATE = "model_update"
    VALIDATION_PROGRESS = "validation_progress"
    ERROR = "error"
    PING = "ping"
    PONG = "pong"

@dataclass
class WSMessage:
    type: MessageType
    data: Dict[str, Any]
    timestamp: str = None
    
    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = datetime.now().isoformat()
    
    def to_json(self) -> str:
        return json.dumps({
            'type': self.type.value,
            'data': self.data,
            'timestamp': self.timestamp
        })

class ConnectionManager:
    def __init__(self):
        # Store active connections by user_id
        self.active_connections: Dict[int, List[WebSocket]] = {}
        # Store job subscriptions (job_id -> set of user_ids)
        self.job_subscriptions: Dict[str, Set[int]] = {}
        # Store user -> jobs mapping for cleanup
        self.user_jobs: Dict[int, Set[str]] = {}
        # Background tasks for each connection
        self.background_tasks: Dict[WebSocket, asyncio.Task] = {}
    
    async def connect(self, websocket: WebSocket, user_id: int) -> None:
        """Accept a new WebSocket connection"""
        await websocket.accept()
        
        # Add to active connections
        if user_id not in self.active_connections:
            self.active_connections[user_id] = []
        self.active_connections[user_id].append(websocket)
        
        # Send connection confirmation
        await self.send_personal_message(
            WSMessage(
                type=MessageType.CONNECTION,
                data={"status": "connected", "user_id": user_id}
            ),
            websocket
        )
        
        # Start ping/pong to keep connection alive
        self.background_tasks[websocket] = asyncio.create_task(
            self._keepalive(websocket, user_id)
        )
        
        logger.info(f"User {user_id} connected via WebSocket")
    
    def disconnect(self, websocket: WebSocket, user_id: int) -> None:
        """Remove a WebSocket connection"""
        if user_id in self.active_connections:
            if websocket in self.active_connections[user_id]:
                self.active_connections[user_id].remove(websocket)
            
            # Clean up if no more connections for this user
            if not self.active_connections[user_id]:
                del self.active_connections[user_id]
                # Clean up job subscriptions
                if user_id in self.user_jobs:
                    for job_id in self.user_jobs[user_id]:
                        if job_id in self.job_subscriptions:
                            self.job_subscriptions[job_id].discard(user_id)
                            if not self.job_subscriptions[job_id]:
                                del self.job_subscriptions[job_id]
                    del self.user_jobs[user_id]
        
        # Cancel background task
        if websocket in self.background_tasks:
            self.background_tasks[websocket].cancel()
            del self.background_tasks[websocket]
        
        logger.info(f"User {user_id} disconnected from WebSocket")
    
    async def subscribe_to_job(self, user_id: int, job_id: str) -> None:
        """Subscribe a user to training job updates"""
        if job_id not in self.job_subscriptions:
            self.job_subscriptions[job_id] = set()
        self.job_subscriptions[job_id].add(user_id)
        
        if user_id not in self.user_jobs:
            self.user_jobs[user_id] = set()
        self.user_jobs[user_id].add(job_id)
        
        logger.info(f"User {user_id} subscribed to job {job_id}")
    
    async def unsubscribe_from_job(self, user_id: int, job_id: str) -> None:
        """Unsubscribe a user from training job updates"""
        if job_id in self.job_subscriptions:
            self.job_subscriptions[job_id].discard(user_id)
            if not self.job_subscriptions[job_id]:
                del self.job_subscriptions[job_id]
        
        if user_id in self.user_jobs:
            self.user_jobs[user_id].discard(job_id)
        
        logger.info(f"User {user_id} unsubscribed from job {job_id}")
    
    async def send_personal_message(self, message: WSMessage, websocket: WebSocket) -> None:
        """Send a message to a specific WebSocket connection"""
        try:
            await websocket.send_text(message.to_json())
        except Exception as e:
            logger.error(f"Error sending message to websocket: {e}")
    
    async def send_user_message(self, message: WSMessage, user_id: int) -> None:
        """Send a message to all connections of a specific user"""
        if user_id in self.active_connections:
            disconnected_sockets = []
            for connection in self.active_connections[user_id]:
                try:
                    await connection.send_text(message.to_json())
                except:
                    disconnected_sockets.append(connection)
            
            # Clean up disconnected sockets
            for socket in disconnected_sockets:
                self.disconnect(socket, user_id)
    
    async def broadcast_job_update(self, job_id: str, message: WSMessage) -> None:
        """Broadcast a message to all users subscribed to a job"""
        if job_id in self.job_subscriptions:
            for user_id in self.job_subscriptions[job_id]:
                await self.send_user_message(message, user_id)
    
    async def broadcast_to_all(self, message: WSMessage) -> None:
        """Broadcast a message to all connected users"""
        for user_id in list(self.active_connections.keys()):
            await self.send_user_message(message, user_id)
    
    async def send_training_progress(self, job_id: str, progress_data: Dict[str, Any]) -> None:
        """Send training progress update to subscribed users"""
        message = WSMessage(
            type=MessageType.TRAINING_PROGRESS,
            data={
                "job_id": job_id,
                **progress_data
            }
        )
        await self.broadcast_job_update(job_id, message)
    
    async def send_training_complete(self, job_id: str, result_data: Dict[str, Any]) -> None:
        """Send training completion notification"""
        message = WSMessage(
            type=MessageType.TRAINING_COMPLETE,
            data={
                "job_id": job_id,
                **result_data
            }
        )
        await self.broadcast_job_update(job_id, message)
    
    async def send_training_failed(self, job_id: str, error_data: Dict[str, Any]) -> None:
        """Send training failure notification"""
        message = WSMessage(
            type=MessageType.TRAINING_FAILED,
            data={
                "job_id": job_id,
                **error_data
            }
        )
        await self.broadcast_job_update(job_id, message)
    
    async def send_validation_progress(self, user_id: int, progress_data: Dict[str, Any]) -> None:
        """Send data validation progress update"""
        message = WSMessage(
            type=MessageType.VALIDATION_PROGRESS,
            data=progress_data
        )
        await self.send_user_message(message, user_id)
    
    async def _keepalive(self, websocket: WebSocket, user_id: int) -> None:
        """Send periodic ping messages to keep connection alive"""
        try:
            while True:
                await asyncio.sleep(30)  # Ping every 30 seconds
                await self.send_personal_message(
                    WSMessage(type=MessageType.PING, data={}),
                    websocket
                )
        except asyncio.CancelledError:
            pass
        except Exception as e:
            logger.error(f"Keepalive error for user {user_id}: {e}")
            self.disconnect(websocket, user_id)
    
    async def handle_message(self, websocket: WebSocket, user_id: int, message: str) -> None:
        """Handle incoming WebSocket messages"""
        try:
            data = json.loads(message)
            message_type = data.get('type')
            
            if message_type == 'subscribe':
                job_id = data.get('job_id')
                if job_id:
                    await self.subscribe_to_job(user_id, job_id)
                    await self.send_personal_message(
                        WSMessage(
                            type=MessageType.CONNECTION,
                            data={"status": "subscribed", "job_id": job_id}
                        ),
                        websocket
                    )
            
            elif message_type == 'unsubscribe':
                job_id = data.get('job_id')
                if job_id:
                    await self.unsubscribe_from_job(user_id, job_id)
                    await self.send_personal_message(
                        WSMessage(
                            type=MessageType.CONNECTION,
                            data={"status": "unsubscribed", "job_id": job_id}
                        ),
                        websocket
                    )
            
            elif message_type == 'pong':
                # Client responded to ping
                pass
            
            else:
                logger.warning(f"Unknown message type: {message_type}")
        
        except json.JSONDecodeError:
            logger.error(f"Invalid JSON message from user {user_id}: {message}")
        except Exception as e:
            logger.error(f"Error handling message from user {user_id}: {e}")

# Global connection manager instance
manager = ConnectionManager()

# WebSocket endpoint handler
async def websocket_endpoint(websocket: WebSocket, user_id: int):
    """Main WebSocket endpoint handler"""
    await manager.connect(websocket, user_id)
    try:
        while True:
            data = await websocket.receive_text()
            await manager.handle_message(websocket, user_id, data)
    except WebSocketDisconnect:
        manager.disconnect(websocket, user_id)
    except Exception as e:
        logger.error(f"WebSocket error for user {user_id}: {e}")
        manager.disconnect(websocket, user_id)

# Integration with training pipeline
class WebSocketProgressReporter:
    """Reporter that sends training progress via WebSocket"""
    
    def __init__(self, job_id: str):
        self.job_id = job_id
    
    async def report_progress(self, percentage: float, status: str, message: str, metrics: Dict[str, Any] = None):
        """Report training progress via WebSocket"""
        await manager.send_training_progress(self.job_id, {
            "progress_percentage": percentage,
            "status": status,
            "message": message,
            "metrics": metrics or {}
        })
    
    async def report_complete(self, result_data: Dict[str, Any]):
        """Report training completion"""
        await manager.send_training_complete(self.job_id, result_data)
    
    async def report_failure(self, error_message: str):
        """Report training failure"""
        await manager.send_training_failed(self.job_id, {
            "error_message": error_message
        })

def get_ws_reporter(job_id: str) -> WebSocketProgressReporter:
    """Get a WebSocket progress reporter for a training job"""
    return WebSocketProgressReporter(job_id)